actor/src/main/scala/org/apache/pekko/pattern/Patterns.scala

```diff
@@ -491,21 +492,50 @@ object Patterns {
    *
    * If attempts are exhausted the returned completion operator is simply the result of invoking attempt.
    * Note that the attempt function will be invoked on the given execution context for subsequent tries
-   * and therefore must be thread safe (not touch unsafe mutable state).
+   * and therefore must be thread safe (i.e. not touch unsafe mutable state).
    */
   def retry[T](attempt: Callable[CompletionStage[T]], attempts: Int, ec: ExecutionContext): CompletionStage[T] = {
     require(attempt != null, "Parameter attempt should not be null.")
     scalaRetry(() => attempt.call().asScala, attempts)(ec).asJava
   }
 
+  /**
+   * Returns an internally retrying [[java.util.concurrent.CompletionStage]].
+   *
+   * When the future is completed, the `shouldRetry` predicate is always been invoked with the result (or `null` if none)
+   * and the exception (or `null` if none). If the `shouldRetry` predicate returns true, then a new attempt is made,
+   * each subsequent attempt will be made after the 'delay' return by `delayFunction` (the input next attempt count start from 1).
+   * Return an empty [[Optional]] instance for no delay.
+   *
+   * If attempts are exhausted the returned completion operator is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries
+   * and therefore must be thread safe (i.e. not touch unsafe mutable state).
+   *
+   * @param attempt       the function to be attempted
+   * @param shouldRetry   the predicate to determine if the attempt should be retried
+   * @param attempts      the maximum number of attempts
+   * @param ec       the execution context
+   * @return the result [[java.util.concurrent.CompletionStage]] which maybe retried
+   *
+   * @since 1.1.0
+   */
+  def retry[T](
+      attempt: Callable[CompletionStage[T]],
+      shouldRetry: BiPredicate[T, Throwable],
+      attempts: Int,
+      ec: ExecutionContext): CompletionStage[T] = {
+    require(attempt != null, "Parameter attempt should not be null.")
+    scalaRetry(() => attempt.call().asScala, (t, e) => shouldRetry.test(t, e), attempts)(ec).asJava
```

Comment:
This class had a lot of converters like `asScala` and `asJava`, it will be great if we could optimized it.

Reply:
Yes, That will need a rewrite Dedicated for Java, that can be done later.